lib/offsite_payments/integrations/poli_pay.rb

```diff
@@ -0,0 +1,247 @@
+          super
+        end
+
+        def credential_based_url
+          options = TransactionBuilder.new(@options, self).to_hash
+          Interface.new(@login, @password).credential_based_url(options)
+        end
+      end
+
+      class TransactionBuilder
+        SUPPORTED_CURRENCIES = ['AUD', 'NZD']
+
+        def initialize(options, helper)
+          @reference        = options.fetch(:order)
+          @timeout          = options[:timeout] # or defaults
+          @success_url      = options.fetch(:success_url, options.fetch(:return_url))
+          @failure_url      = options.fetch(:failure_url, options.fetch(:return_url))
+          @notification_url = helper.notify_url
+          @homepage_url     = options.fetch(:homepage_url)
+          self.amount       = options.fetch(:amount)
+          self.currency     = options.fetch(:currency)
+        end
+
+        def amount=(amount)
+          @amount = '%.2f' % amount.to_f.round(2)
+        end
+
+        def currency=(currency)
+          unless SUPPORTED_CURRENCIES.include?(currency)
+            raise ArgumentError, "Unsupported currency"
+          end
+          @currency = currency
+        end
+
+        def current_time
+          Time.current.utc.strftime("%Y-%m-%dT%H:%M:%S")
+        end
+
+        def to_hash
+          {
+            Amount:              @amount,
+            CurrencyCode:        @currency,
+            MerchantReference:   @reference,
+            SuccessURL:          @success_url,
+            FailureURL:          @failure_url,
+            NotificationUrl:     @notification_url,
+            MerchantHomepageURL: @homepage_url,
+            Timeout:             @timeout,
+            MerchantDateTime:    current_time
+          }.reject { |_, v| v.blank? }
+        end
+      end
+
+      class Notification < OffsitePayments::Notification
+        def initialize(params, options = {})
+          token = params.fetch('token')
+          @params = Interface.new(options[:login], options[:password]).query(token)
+        end
+
+        def acknowledge
+          true # always valid as we fetch direct from poli
+        end
+
+        def complete?
+          @params['TransactionStatusCode'] == 'Completed'
+        end
+
+        def success?
+          complete? && gross && gross > 0
+        end
+
+        def gross
+          @params['AmountPaid']
+        end
+
+        def currency
+          @params['CurrencyCode']
+        end
+
+        def order_id
+          @params['MerchantReference']
+        end
+
+        def transaction_id
+          @params['TransactionRefNo']
+        end
+
+        def message # only when error
+          @params['ErrorMessage']
+        end
+      end
+
+      class Return < OffsitePayments::Return
+        def initialize(query_string, options={})
+          @notification = Notification.new(query_string, options)
+        end
+      end
+
+      class Error < StandardError
```

Comment:
It was meant to be a superclass so you can easily rescue all PoliPay 
errors, of course it's effectively the same as RequestError because that 
is the only subclass for now, but for the API's sake it is nice to be 
able to say "is it a PoliPay::Error?" and have that work into the future.

On 12/01/16 13:09, Alessandro Berardi wrote:
>
> In lib/offsite_payments/integrations/poli_pay.rb 
> <https://github.com/sealink/offsite_payments/pull/1#discussion_r49409791>:
>
> > +        def transaction_id
> > +          @params['TransactionRefNo']
> > +        end
> > +
> > +        def message # only when error
> > +          @params['ErrorMessage']
> > +        end
> > +      end
> > +
> > +      class Return < OffsitePayments::Return
> > +        def initialize(query_string, options={})
> > +          @notification = Notification.new(query_string, options)
> > +        end
> > +      end
> > +
> > +      class Error < StandardError
>
> What's the point of this guy really?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/sealink/offsite_payments/pull/1/files#r49409791>.
>
>
> ______________________________________________________________________
> This email has been scanned by the Symantec Email Security.cloud service.
> For more information please visit http://www.symanteccloud.com
> ______________________________________________________________________